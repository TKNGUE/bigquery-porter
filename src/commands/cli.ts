import * as fs from 'node:fs';

import { formatLocalfiles } from '../../src/commands/fix.js';
import { PushContext, pushLocalFilesToBigQuery } from '../../src/commands/push.js';
import { createBundleSQL } from '../../src/commands/bundle.js';
import { pullBigQueryResources } from '../../src/commands/pull.js';
import { importBigQueryResources } from '../../src/commands/import.js';

import { buildThrottledBigQueryClient } from '../../src/bigquery.js';
import type { Query } from '@google-cloud/bigquery';

import 'process';
import { Command } from 'commander';

export function createCLI() {
  const program = new Command();

  program
    .description('Easy and Quick BigQuery Deployment Tool')
    // Global Options
    .option('-n, --threads <threads>', 'API Call Concurrency', '8')
    .option('-C, --root-path <rootPath>', 'Root Directory', './bigquery')
    .option(
      '--enable-datalineage',
      `Enable Data Lineage integration to resolve DAG dependency`,
      false,
    )
    .option(
      '--format <reporter>',
      'formatter option: console, json',
      'console',
    );

  const pushCommand = new Command('push')
    .description(
      'Deploy your local BigQuery Resources in topological-sorted order',
    )
    .argument('[projects...]')
    .option(
      '--sweeping-mode <mode>',
      'Strategy for undefined/orphans BigQuery resources. option: confirm(default), ignore, rename',
      'confirm',
    )
    .option(
      '--label <key:value...>',
      'A label to set on a query job. The format is "key:value"; repeat this option to specify a list of values',
    )
    .option(
      '-p, --parameter <key:value...>',
      `Either a file containing a JSON list of query parameters, or a query parameter in the form "name:type:value". ` +
      `An empty name produces a positional parameter. The type may be omitted to assume STRING: name::value or ::value. ` +
      `The value "NULL" produces a null value. repeat this option to specify a list of values`,
    )
    .option(
      '--maximum_bytes_billed <number of bytes>',
      'The upper limit of bytes billed for the query.',
    )
    .option('--dry-run', 'Dry Run', false)
    .action(async (cmdProjects: string[] | undefined, _, cmd) => {
      const cmdOptions = cmd.optsWithGlobals();
      const projects = cmdProjects && cmdProjects.length > 0
        ? cmdProjects
        : ['@default'];

      const rootDir = cmdOptions.rootPath;
      if (!rootDir) {
        console.error('CLI Error');
        return;
      }

      const jobOption: Query = {};
      jobOption.dryRun = cmdOptions.dryRun ?? false;
      if (cmdOptions.maximumBytesBilled) {
        jobOption.maximumBytesBilled = cmdOptions.maximumBytesBilled;
      }

      jobOption.labels = { 'bqporter-enable': 'true' };
      if (cmdOptions.label) {
        jobOption.labels = {
          ...jobOption.labels,
          ...Object.fromEntries(
            (cmdOptions.label as string[]).map((l) => l.split(':')),
          ),
        };
      }

      if (cmdOptions.parameter) {
        jobOption.params = {} as { [param: string]: any };
        for (const param of cmdOptions.parameter as string[]) {
          if (fs.existsSync(param) && fs.lstatSync(param).isFile()) {
            jobOption.params = {
              ...jobOption.params,
              ...JSON.parse(fs.readFileSync(param, 'utf8')),
            };
            continue;
          }

          const elms = param.split(':');
          const name = elms[0];
          if (name === undefined) {
            console.error(`Invalid Parameter: ${param}`);
            return;
          }
          const rawValue = elms[2];
          if (rawValue === undefined) {
            console.error(`Invalid Parameter: ${param}`);
            return;
          }

          (jobOption.params as { [param: string]: any })[name] = (() => {
            if (!elms[1]) {
              return rawValue;
            }

            if (elms[1]?.toUpperCase() === 'STRING') {
              return rawValue;
            }

            if (elms[1]?.toUpperCase() === 'INTGER') {
              return parseInt(rawValue);
            }

            if (elms[1]?.toUpperCase() === 'BOOL') {
              return rawValue.toLowerCase() == 'true';
            }

            if (elms[1]?.toUpperCase() === 'TIMESTAMP') {
              const p = new Date(Date.parse(rawValue));
              if (!isNaN(p.valueOf())) {
                return p;
              }
            }

            if (elms[1]) {
              try {
                return JSON.parse(rawValue);
              } catch (e) {
                return rawValue;
              }
            }
            return rawValue;
          })();
        }
      }

      const bqClient = buildThrottledBigQueryClient(
        parseInt(cmdOptions.threads),
        500,
      );

      const sweepMode: PushContext['SweepStrategy']['mode'] = (() => {
        if(!cmdOptions.sweepingMode) {
          return 'confirm'
        }
        switch (cmdOptions.sweepingMode) {
          case 'ignore':
          case 'rename_and_7d_expire':
          case 'confirm':
          case 'force':
            return cmdOptions.sweepingMode
          default:
            // Error for invalid option
            throw Error(`Invalid --sweeping-mode option: ${cmdOptions.sweepingMode}`);
        }
      })();

      for (const project of projects) {
        const ctx = {
          BigQuery: {
            client: bqClient,
            projectId: project,
          },
          rootPath: rootDir,
          dryRun: cmdOptions.dryRun ?? false,
          SweepStrategy: {
            mode: sweepMode,
            ignorePrefix: 'zorphan__',
          },
          reporter: cmdOptions.format ?? 'console',
          enableDataLineage: cmdOptions.enableDatalineage ?? false,
        } as PushContext;

        const failed = await pushLocalFilesToBigQuery(ctx, jobOption);
        if (failed > 0) {
          program.error(`Some tasks are failed(Failed: ${failed})`, {
            exitCode: 1,
          });
        }
      }
    });

  const pullCommand = new Command('pull')
    .description('pull dataset and its tabald and routine information')
    .argument('[ids...]')
    .option('--all', 'Pulling All BugQuery Datasets', false)
    .option('--with-ddl', 'Pulling BigQuery Resources with DDL SQL', false)
    .action(async (cmdBQIDs: string[] | undefined, _, cmd) => {
      const cmdOptions = cmd.optsWithGlobals();
      const BQIDs = cmdBQIDs && cmdBQIDs.length > 0 ? cmdBQIDs : ['@default'];

      const failed = await (async () => {
        return await pullBigQueryResources({
          BQIDs,
          BigQuery: buildThrottledBigQueryClient(20, 500),
          rootPath: cmdOptions.rootPath,
          withDDL: cmdOptions.withDdl,
          forceAll: cmdOptions.all,
          // concurrency: cmdOptions.concurrency,
          reporter: cmdOptions.format ?? 'console',
        });
      })();

      if (failed > 0) {
        program.error(`Some tasks are failed(Failed: ${failed})`, {
          exitCode: 1,
        });
      }
    });

  const formatCommmand = new Command('format')
    .description('Fix reference in local DDL files')
    .option('--dry-run', 'dry run', false)
    .action(async (_, cmd) => {
      const cmdOptions = cmd.optsWithGlobals();
      const options = {
        dryRun: cmdOptions.dryRun,
      };
      await formatLocalfiles(cmdOptions.rootPath ?? './bigquery/', options);
    });

  const bundleCommand = new Command('bundle')
    .description(
      'Bundle SQLs into single execuatale BigQuery Script',
    )
    .argument('[projects...]')
    .action(async (cmdProjects: string[] | undefined, _, cmd) => {
      const cmdOptions = cmd.optsWithGlobals();
      const projects = cmdProjects && cmdProjects.length > 0
        ? cmdProjects
        : ['@default'];

      const rootDir = cmdOptions.rootPath;
      if (!rootDir) {
        console.error('CLI Error');
        return;
      }

      const bqClient = buildThrottledBigQueryClient(
        parseInt(cmdOptions.threads),
        500,
      );
      for (const project of projects) {
        const ctx = {
          BigQuery: {
            client: bqClient,
            projectId: project ?? '@default',
          },
          rootPath: rootDir,
          dryRun: cmdOptions.dryRun ?? false,
          force: cmdOptions.force ?? false,
          reporter: cmdOptions.format ?? 'console',
          enableDataLineage: cmdOptions.enableDatalineage ?? false,
        };

        const sql = await createBundleSQL(ctx);
        console.log(sql);
      }
    });

  const importCommand = new Command('import')
    .description(
      'Import other dataset UDF into specified dataset',
    )
    .argument('<destination>')
    .argument('[targets...]')
    .action(
      async (destination: string, cmdTargets: string[] | undefined, _, cmd) => {
        const cmdOptions = cmd.optsWithGlobals();
        const rootDir = cmdOptions.rootPath;
        if (!rootDir) {
          console.error('CLI Error');
          return;
        }

        const bqClient = buildThrottledBigQueryClient(
          parseInt(cmdOptions.threads),
          500,
        );

        // Parse targets 'bqutils.fn.sure_nonnull' into [{'project': 'bqutils', dataset: 'fn', routine_id: 'sure_nonnull'}]
        const targets = cmdTargets?.map((target) => {
          const elms = target.split('.');
          if (elms.length !== 3) {
            throw new Error(`Invalid target: ${target}`);
          }
          return {
            project: elms[0] as string,
            dataset: elms[1] as string,
            routine_id: elms[2] as string,
          };
        }) ?? [];

        let paramDestination: { project: string; dataset: string };
        const [projectOrDataset, destinationDataset] = destination.split('.');
        if (destinationDataset) {
          paramDestination = {
            project: destinationDataset,
            dataset: destinationDataset,
          };
        } else if (projectOrDataset) {
          paramDestination = {
            project: '@default',
            dataset: projectOrDataset,
          };
        } else {
          throw new Error(`Invalid destination: ${destination}`);
        }

        const ctx = {
          bigQuery: bqClient,
          rootPath: rootDir,
          destination: paramDestination,
          importTargets: targets,
          options: {
            is_update: true,
          },
        };

        await importBigQueryResources(ctx);
      },
    );

  program.addCommand(pushCommand);
  program.addCommand(pullCommand);
  program.addCommand(formatCommmand);
  program.addCommand(bundleCommand);
  program.addCommand(importCommand);

  return program;
}
