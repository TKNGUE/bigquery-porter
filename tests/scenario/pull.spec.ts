import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest';

import { tmpdir } from 'node:os';
import * as path from 'node:path';
import * as fs from 'node:fs';

import { Command } from 'commander';

import { createCLI } from '../../src/commands/cli.js';
import { walk } from '../../src/util.js';

interface CLITestContext {
  cli: Command;
  out: string[];
  err: string[];
  rootPath: string;
}

describe('CLIv2: pull', () => {
  const tempDir = path.join(tmpdir(), `bqport-`);
  const setupCommanderForTest = (c: Command, out: string[], err: string[]) => {
    c
      .exitOverride((e: Error) => {
        throw e;
      })
      .configureOutput({
        writeOut: (s) => out.push(s),
        writeErr: (s) => err.push(s),
      });
  };

  const crawlFs = async (root: string) => {
    const file2content: Map<string, string> = new Map();
    for (const f of await walk(root)) {
      const content = await fs.promises.readFile(f, 'utf-8');
      file2content.set(path.relative(root, f), content);
    }
    return file2content;
  };

  beforeEach<CLITestContext>(async (ctx) => {
    const out: string[] = [];
    const err: string[] = [];
    const cli = createCLI();
    setupCommanderForTest(cli, out, err);
    for (const c of cli.commands) {
      setupCommanderForTest(c, out, err);
    }

    vi.spyOn(console, 'log')
      .mockImplementation((s: string) => {
        out.push(s);
      });
    vi.spyOn(console, 'error')
      .mockImplementation((s: string) => {
        err.push(s);
      });

    ctx.cli = cli;
    ctx.out = out;
    ctx.err = err;
    ctx.rootPath = path.resolve(fs.mkdtempSync(tempDir));
  });

  afterEach(async () => {
    vi.resetAllMocks();
  });

  it<CLITestContext>(
    `pull --help`,
    async ({ meta, cli, out, err, rootPath }) => {
      await expect(
        async () => {
          await cli.parseAsync(meta.name.split(' '), { from: 'user' });
        },
      ).rejects.toThrow();
      expect(await crawlFs(rootPath)).toMatchSnapshot();
      expect(err).toMatchSnapshot();
      expect(out).toMatchSnapshot();
    },
  );

  it<CLITestContext>(
    `pull --all`,
    async ({ meta, cli, err, rootPath }) => {
      await cli.parseAsync([...meta.name.split(' '), ...['-C', rootPath]], {
        from: 'user',
      });
      expect(await crawlFs(path.join(rootPath)))
        .toMatchSnapshot();
      expect(err).toMatchSnapshot();
      // expect(out).toMatchSnapshot();
    },
  );

  it<CLITestContext>(
    `pull --format=json bigquery-public-data.baseball`,
    async ({ meta, cli, out, err, rootPath }) => {
      await cli.parseAsync([...meta.name.split(' '), ...['-C', rootPath]], {
        from: 'user',
      });
      expect(new Set(await crawlFs(path.join(rootPath))))
        .toMatchSnapshot();
      expect(err).toMatchSnapshot();
      expect(new Set(out)).toMatchSnapshot();
      expect(out.length).toMatchSnapshot();
    },
  );

  it<CLITestContext>(
    `pull --format=json --with-ddl bigquery-public-data.baseball`,
    async ({ meta, cli, out, err, rootPath }) => {
      await cli.parseAsync([...meta.name.split(' '), ...['-C', rootPath]], {
        from: 'user',
      });
      expect(new Set(await crawlFs(path.join(rootPath))))
        .toMatchSnapshot();
      expect(err).toMatchSnapshot();
      expect(new Set(out)).toMatchSnapshot();
      expect(out.length).toMatchSnapshot();
    },
  );
});
