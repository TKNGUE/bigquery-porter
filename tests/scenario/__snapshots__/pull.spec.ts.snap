// Vitest Snapshot v1

exports[`CLIv2: pull > pull --all 1`] = `
Map {
  "sandbox/@models/mymodel/metadata.json" => "{
  \\"modelType\\": \\"LINEAR_REGRESSION\\",
  \\"featureColumns\\": [
    {
      \\"name\\": \\"column1\\",
      \\"type\\": {
        \\"typeKind\\": \\"INT64\\"
      }
    },
    {
      \\"name\\": \\"column2\\",
      \\"type\\": {
        \\"typeKind\\": \\"INT64\\"
      }
    },
    {
      \\"name\\": \\"column3\\",
      \\"type\\": {
        \\"typeKind\\": \\"INT64\\"
      }
    }
  ],
  \\"labelColumns\\": [
    {
      \\"name\\": \\"predicted_label\\",
      \\"type\\": {
        \\"typeKind\\": \\"INT64\\"
      }
    }
  ],
  \\"trainingRuns\\": [
    {
      \\"trainingOptions\\": {
        \\"maxIterations\\": \\"5\\",
        \\"lossType\\": \\"MEAN_SQUARED_LOSS\\",
        \\"l1Regularization\\": 1,
        \\"l2Regularization\\": 0,
        \\"minRelativeProgress\\": 0.01,
        \\"warmStart\\": false,
        \\"earlyStop\\": true,
        \\"dataSplitMethod\\": \\"AUTO_SPLIT\\",
        \\"learnRateStrategy\\": \\"LINE_SEARCH\\",
        \\"initialLearnRate\\": 0.15,
        \\"optimizationStrategy\\": \\"BATCH_GRADIENT_DESCENT\\",
        \\"calculatePValues\\": false,
        \\"preserveInputStructs\\": false
      },
      \\"trainingStartTime\\": \\"1672534389200\\",
      \\"results\\": [
        {
          \\"index\\": 0,
          \\"durationMs\\": \\"2231\\",
          \\"trainingLoss\\": 1491.053977063737,
          \\"learnRate\\": 0.3
        },
        {
          \\"index\\": 1,
          \\"durationMs\\": \\"1991\\",
          \\"trainingLoss\\": 241.92978201881198,
          \\"learnRate\\": 0.6
        },
        {
          \\"index\\": 2,
          \\"durationMs\\": \\"2178\\",
          \\"trainingLoss\\": 41.69727113362372,
          \\"learnRate\\": 0.6
        },
        {
          \\"index\\": 3,
          \\"durationMs\\": \\"2166\\",
          \\"trainingLoss\\": 7.920357012251959,
          \\"learnRate\\": 0.6
        },
        {
          \\"index\\": 4,
          \\"durationMs\\": \\"2040\\",
          \\"trainingLoss\\": 2.6036132184550516,
          \\"learnRate\\": 0.6
        }
      ],
      \\"evaluationMetrics\\": {
        \\"regressionMetrics\\": {
          \\"meanAbsoluteError\\": 1.3283107802514804,
          \\"meanSquaredError\\": 2.6036132184550516,
          \\"meanSquaredLogError\\": 0.0012176719511574185,
          \\"medianAbsoluteError\\": 1.0159802433960294,
          \\"rSquared\\": 0.9968441051897514
        }
      },
      \\"startTime\\": \\"2023-01-01T00:53:09.200Z\\"
    }
  ]
}",
  "sandbox/@routines/sample_function/README.md" => "test",
  "sandbox/@routines/sample_function/metadata.json" => "{
  \\"routineType\\": \\"SCALAR_FUNCTION\\",
  \\"language\\": \\"SQL\\",
  \\"arguments\\": [
    {
      \\"name\\": \\"argument\\",
      \\"dataType\\": {
        \\"typeKind\\": \\"INT64\\"
      }
    }
  ],
  \\"returnType\\": {
    \\"typeKind\\": \\"INT64\\"
  }
}",
  "sandbox/@routines/sample_procedure/metadata.json" => "{
  \\"routineType\\": \\"PROCEDURE\\",
  \\"language\\": \\"SQL\\",
  \\"arguments\\": [
    {
      \\"name\\": \\"sample_argument\\",
      \\"mode\\": \\"IN\\",
      \\"dataType\\": {
        \\"typeKind\\": \\"INT64\\"
      }
    }
  ]
}",
  "sandbox/@routines/sample_tvf/metadata.json" => "{
  \\"routineType\\": \\"TABLE_VALUED_FUNCTION\\",
  \\"language\\": \\"SQL\\",
  \\"arguments\\": [
    {
      \\"name\\": \\"argument\\",
      \\"dataType\\": {
        \\"typeKind\\": \\"INT64\\"
      }
    }
  ],
  \\"returnTableType\\": {
    \\"columns\\": [
      {
        \\"name\\": \\"a\\",
        \\"type\\": {
          \\"typeKind\\": \\"INT64\\"
        }
      }
    ]
  }
}",
  "sandbox/sample_clone_table/metadata.json" => "{
  \\"type\\": \\"TABLE\\",
  \\"clustering\\": {
    \\"fields\\": [
      \\"a\\"
    ]
  }
}",
  "sandbox/sample_clone_table/schema.json" => "[
  {
    \\"name\\": \\"a\\",
    \\"type\\": \\"INTEGER\\"
  }
]",
  "sandbox/sample_materialized_view/metadata.json" => "{
  \\"type\\": \\"MATERIALIZED_VIEW\\"
}",
  "sandbox/sample_materialized_view/schema.json" => "[
  {
    \\"name\\": \\"a\\",
    \\"type\\": \\"INTEGER\\"
  }
]",
  "sandbox/sample_partition_table/metadata.json" => "{
  \\"type\\": \\"TABLE\\",
  \\"timePartitioning\\": {
    \\"type\\": \\"DAY\\",
    \\"field\\": \\"date_jst\\"
  }
}",
  "sandbox/sample_partition_table/schema.json" => "[
  {
    \\"name\\": \\"date_jst\\",
    \\"type\\": \\"DATE\\"
  }
]",
  "sandbox/sample_sharding_@/metadata.json" => "{
  \\"type\\": \\"TABLE\\"
}",
  "sandbox/sample_sharding_@/schema.json" => "[
  {
    \\"name\\": \\"date_jst\\",
    \\"type\\": \\"DATE\\"
  }
]",
  "sandbox/sample_snapshot_table/metadata.json" => "{
  \\"type\\": \\"SNAPSHOT\\",
  \\"snapshotDefinition\\": {
    \\"baseTableReference\\": {
      \\"projectId\\": \\"takegue\\",
      \\"datasetId\\": \\"sandbox\\",
      \\"tableId\\": \\"sample_table\\"
    },
    \\"snapshotTime\\": \\"2023-01-01T01:12:07.322Z\\"
  },
  \\"cloneDefinition\\": {
    \\"baseTableReference\\": {
      \\"projectId\\": \\"takegue\\",
      \\"datasetId\\": \\"sandbox\\",
      \\"tableId\\": \\"sample_table\\"
    },
    \\"snapshotTime\\": \\"2023-01-01T01:12:07.322Z\\"
  }
}",
  "sandbox/sample_snapshot_table/schema.json" => "[
  {
    \\"name\\": \\"a\\",
    \\"type\\": \\"INTEGER\\",
    \\"description\\": \\"ふがふが\\\\nふが\\"
  }
]",
  "sandbox/sample_table/metadata.json" => "{
  \\"type\\": \\"TABLE\\",
  \\"clustering\\": {
    \\"fields\\": [
      \\"a\\"
    ]
  }
}",
  "sandbox/sample_table/schema.json" => "[
  {
    \\"name\\": \\"a\\",
    \\"type\\": \\"INTEGER\\",
    \\"description\\": \\"ふがふが\\\\nふが\\"
  }
]",
  "sandbox/sample_view/metadata.json" => "{
  \\"type\\": \\"VIEW\\"
}",
  "sandbox/sample_view/schema.json" => "[
  {
    \\"name\\": \\"unique_key\\",
    \\"type\\": \\"STRING\\",
    \\"mode\\": \\"NULLABLE\\"
  },
  {
    \\"name\\": \\"complaint_description\\",
    \\"type\\": \\"STRING\\",
    \\"mode\\": \\"NULLABLE\\"
  },
  {
    \\"name\\": \\"source\\",
    \\"type\\": \\"STRING\\",
    \\"mode\\": \\"NULLABLE\\"
  },
  {
    \\"name\\": \\"status\\",
    \\"type\\": \\"STRING\\",
    \\"mode\\": \\"NULLABLE\\"
  },
  {
    \\"name\\": \\"status_change_date\\",
    \\"type\\": \\"TIMESTAMP\\",
    \\"mode\\": \\"NULLABLE\\"
  },
  {
    \\"name\\": \\"created_date\\",
    \\"type\\": \\"TIMESTAMP\\",
    \\"mode\\": \\"NULLABLE\\"
  },
  {
    \\"name\\": \\"last_update_date\\",
    \\"type\\": \\"TIMESTAMP\\",
    \\"mode\\": \\"NULLABLE\\"
  },
  {
    \\"name\\": \\"close_date\\",
    \\"type\\": \\"TIMESTAMP\\",
    \\"mode\\": \\"NULLABLE\\"
  },
  {
    \\"name\\": \\"incident_address\\",
    \\"type\\": \\"STRING\\",
    \\"mode\\": \\"NULLABLE\\"
  },
  {
    \\"name\\": \\"street_number\\",
    \\"type\\": \\"STRING\\",
    \\"mode\\": \\"NULLABLE\\"
  },
  {
    \\"name\\": \\"street_name\\",
    \\"type\\": \\"STRING\\",
    \\"mode\\": \\"NULLABLE\\"
  },
  {
    \\"name\\": \\"city\\",
    \\"type\\": \\"STRING\\",
    \\"mode\\": \\"NULLABLE\\"
  },
  {
    \\"name\\": \\"incident_zip\\",
    \\"type\\": \\"INTEGER\\",
    \\"mode\\": \\"NULLABLE\\"
  },
  {
    \\"name\\": \\"county\\",
    \\"type\\": \\"STRING\\",
    \\"mode\\": \\"NULLABLE\\"
  },
  {
    \\"name\\": \\"state_plane_x_coordinate\\",
    \\"type\\": \\"STRING\\",
    \\"mode\\": \\"NULLABLE\\"
  },
  {
    \\"name\\": \\"state_plane_y_coordinate\\",
    \\"type\\": \\"FLOAT\\",
    \\"mode\\": \\"NULLABLE\\"
  },
  {
    \\"name\\": \\"latitude\\",
    \\"type\\": \\"FLOAT\\",
    \\"mode\\": \\"NULLABLE\\"
  },
  {
    \\"name\\": \\"longitude\\",
    \\"type\\": \\"FLOAT\\",
    \\"mode\\": \\"NULLABLE\\"
  },
  {
    \\"name\\": \\"location\\",
    \\"type\\": \\"STRING\\",
    \\"mode\\": \\"NULLABLE\\"
  },
  {
    \\"name\\": \\"council_district_code\\",
    \\"type\\": \\"INTEGER\\",
    \\"mode\\": \\"NULLABLE\\"
  },
  {
    \\"name\\": \\"map_page\\",
    \\"type\\": \\"STRING\\",
    \\"mode\\": \\"NULLABLE\\"
  },
  {
    \\"name\\": \\"map_tile\\",
    \\"type\\": \\"STRING\\",
    \\"mode\\": \\"NULLABLE\\"
  }
]",
  "sandbox/sample_view/view.sql" => "SELECT * FROM \`bigquery-public-data.austin_311.311_service_requests\` LIMIT 1000
",
  "zzHXMX7Z/@routines/sample_routine/metadata.json" => "{
  \\"routineType\\": \\"SCALAR_FUNCTION\\",
  \\"language\\": \\"SQL\\",
  \\"arguments\\": [
    {
      \\"name\\": \\"from\\",
      \\"dataType\\": {
        \\"typeKind\\": \\"STRING\\"
      }
    }
  ],
  \\"returnType\\": {
    \\"typeKind\\": \\"STRING\\"
  }
}",
  "zzHXMX7Z/sample_table/README.md" => "sample table",
  "zzHXMX7Z/sample_table/metadata.json" => "{
  \\"type\\": \\"TABLE\\"
}",
  "zzHXMX7Z/sample_table/schema.json" => "[
  {
    \\"name\\": \\"id\\",
    \\"type\\": \\"INTEGER\\"
  },
  {
    \\"name\\": \\"name\\",
    \\"type\\": \\"STRING\\"
  }
]",
  "zzrIxASZ/@routines/sample_routine/metadata.json" => "{
  \\"routineType\\": \\"SCALAR_FUNCTION\\",
  \\"language\\": \\"SQL\\",
  \\"arguments\\": [
    {
      \\"name\\": \\"from\\",
      \\"dataType\\": {
        \\"typeKind\\": \\"STRING\\"
      }
    }
  ],
  \\"returnType\\": {
    \\"typeKind\\": \\"STRING\\"
  }
}",
  "zzrIxASZ/sample_table/README.md" => "sample table",
  "zzrIxASZ/sample_table/metadata.json" => "{
  \\"type\\": \\"TABLE\\"
}",
  "zzrIxASZ/sample_table/schema.json" => "[
  {
    \\"name\\": \\"id\\",
    \\"type\\": \\"INTEGER\\"
  },
  {
    \\"name\\": \\"name\\",
    \\"type\\": \\"STRING\\"
  }
]",
}
`;

exports[`CLIv2: pull > pull --all 2`] = `
[
  "/tmp/bqport-kYtOIQ/@default/sandbox/@models/mymodel",
  "/tmp/bqport-kYtOIQ/@default/sandbox/@routines/sample_function",
  "/tmp/bqport-kYtOIQ/@default/sandbox/@routines/sample_procedure",
  "/tmp/bqport-kYtOIQ/@default/sandbox/@routines/sample_tvf",
  "/tmp/bqport-kYtOIQ/@default/zzrIxASZ/@routines/sample_routine",
  "/tmp/bqport-kYtOIQ/@default/zzHXMX7Z/@routines/sample_routine",
  "/tmp/bqport-kYtOIQ/@default/sandbox/sample_clone_table",
  "/tmp/bqport-kYtOIQ/@default/sandbox/sample_materialized_view",
  "/tmp/bqport-kYtOIQ/@default/sandbox/sample_partition_table",
  "/tmp/bqport-kYtOIQ/@default/sandbox/sample_sharding_@",
  "/tmp/bqport-kYtOIQ/@default/sandbox/sample_snapshot_table",
  "/tmp/bqport-kYtOIQ/@default/sandbox/sample_table",
  "/tmp/bqport-kYtOIQ/@default/sandbox/sample_view",
  "/tmp/bqport-kYtOIQ/@default/zzrIxASZ/sample_table",
  "/tmp/bqport-kYtOIQ/@default/zzHXMX7Z/sample_table",
]
`;

exports[`CLIv2: pull > pull --format=json bigquery-public-data.baseball 1`] = `Map {}`;

exports[`CLIv2: pull > pull --format=json bigquery-public-data.baseball 2`] = `[]`;

exports[`CLIv2: pull > pull --format=json bigquery-public-data.baseball 3`] = `
[
  "{\\"name\\":\\"# Check All Dataset and Resources\\",\\"result\\":\\"Total 1 resources\\"}",
]
`;

exports[`CLIv2: pull > pull --help 1`] = `Map {}`;

exports[`CLIv2: pull > pull --help 2`] = `[]`;

exports[`CLIv2: pull > pull --help 3`] = `
[
  "Usage: program pull [options] [ids...]

pull dataset and its tabald and routine information

Options:
  --all       Pulling All BugQuery Datasets (default: false)
  --with-ddl  Pulling BigQuery Resources with DDL SQL (default: false)
  -h, --help  display help for command
",
]
`;
